### ISSUE TITLE
npm install yarn --global fails in docker container

### ISSUE THREAD

**Issue by jakabk** [yarnpkg/yarn#8358]
### Bug description
I've tried to install the latest yarn globally using lts/carbon npm.
It's failed with some permission denied message.
As I see the version which has been tried to install is not released: 1.22.7.
I haven't got problems with 1.22.5 but npm install yarn --global brakes our CI/CD pipeline.
The installation was tried in a docker container as root.

**Command**

```sh
nvm use lts/carbon
npm install yarn --global
```

**What is the current behavior?**

```
root@5ebd9bbf467d:/# npm install yarn --global

> yarn@1.22.7 preinstall /root/.nvm/versions/node/v8.17.0/lib/node_modules/yarn
> node ./preinstall.js

sh: 1: node: Permission denied
npm ERR! code ELIFECYCLE
npm ERR! syscall spawn
npm ERR! file sh
npm ERR! errno ENOENT
npm ERR! yarn@1.22.7 preinstall: `node ./preinstall.js`
npm ERR! spawn ENOENT
npm ERR! 
npm ERR! Failed at the yarn@1.22.7 preinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.

npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-09-28T10_46_23_638Z-debug.log
root@5ebd9bbf467d:/# 
```

**What is the expected behavior?**

```
root@5ebd9bbf467d:/# npm install yarn@1.22.5 --global
/root/.nvm/versions/node/v8.17.0/bin/yarn -> /root/.nvm/versions/node/v8.17.0/lib/node_modules/yarn/bin/yarn.js
/root/.nvm/versions/node/v8.17.0/bin/yarnpkg -> /root/.nvm/versions/node/v8.17.0/lib/node_modules/yarn/bin/yarn.js
+ yarn@1.22.5
added 1 package in 0.345s
```

**Steps to Reproduce**

1. [install nvm](https://github.com/nvm-sh/nvm#install--update-script)
2. nvm install lts/carbon
3. nvm use lts/carbon

**Environment**

- Node Version: `v8.17.0`
- npm version: `6.13.4`
- Yarn v1 Version: `1.22.7` 
- OS and version: `Debian 10 in docker container`

**Comment by dcombrouse** [comment]
Same issue since version `1.22.6` but I don't have error log

On [npm](https://www.npmjs.com/package/yarn/v/1.22.6) there are versions 1.22.6 and 1.22.7 but on git I don't see these [release](https://github.com/yarnpkg/yarn/releases)
They are deleted from git repo ?

**Comment by arcanis** [comment]
We've added a preinstall script in the 1.22.6. Given that they need to ship outside of the standard release pipeline which is broken, they haven't generated formal GH releases (I added them to the changelog tho).

Can you provide more details? In particular, can you explain why your setup cannot execute the `node` binary? If I get a better understanding of what happens I can implement a fix.

```
sh: 1: node: Permission denied
```

**Comment by arcanis** [comment]
From what I can see (https://github.com/nvm-sh/nvm/issues/1407), the problem is that you're installing nvm as root, and when npm runs the lifecycle scripts it downgrades the permissions, making it impossible to even run Node scripts, since the `node` binary can only be accessed by `root`.

I'll make the script tolerant to such failures, but you might want to look at this, as running npm as root is perhaps not recommended.

**Comment by arcanis** [comment]
I've just released the 1.22.8 which uses the following script; it should prevent permission errors (since the script aims to make a future migration easier it's not a critical problem if it can't run), at the cost of the script not running anymore on Windows. I'll update the changelog.

```
:; (node ./preinstall.js >& /dev/null || true)
```

**Comment by hvdklauw** [comment]
We just did a deploy and the new version (1.22.8) got pulled in.. it gave us the following error:
```
$ npm install -g yarn
> yarn@1.22.8 preinstall /usr/lib/node_modules/yarn
> :; (node ./preinstall.js >& /dev/null || true)
sh: 1: Syntax error: Bad fd number
npm ERR! code ELIFECYCLE
npm ERR! errno 2
npm ERR! yarn@1.22.8 preinstall: `:; (node ./preinstall.js >& /dev/null || true)`
npm ERR! Exit status 2
npm ERR! 
npm ERR! Failed at the yarn@1.22.8 preinstall script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
npm ERR! A complete log of this run can be found in:
npm ERR!     /root/.npm/_logs/2020-09-28T13_08_05_779Z-debug.log
```

**Comment by hvdklauw** [comment]
That should be 
```
:; (node ./preinstall.js 2>&1 /dev/null || true)
```
I think

**Comment by gjvoosten** [comment]
@arcanis Can this be fixed real soon (v1.22.9)? It is blocking our work…

**Comment by gjvoosten** [comment]
> That should be
> 
> ```
> :; (node ./preinstall.js 2>&1 /dev/null || true)
> ```
> 
> I think

Probably more like:
```
:; (node ./preinstall.js > /dev/null 2>&1 || true)
```
if the intention is to discard both stdout and stderr.

**Comment by dcombrouse** [comment]
Same for me but I don't have this trace 

``` bash
root@bee39e307c36:/tmp# npm install -g yarn --verbose
npm info it worked if it ends with ok
npm verb cli [ '/usr/bin/node',
npm verb cli   '/usr/bin/npm',
npm verb cli   'install',
npm verb cli   '-g',
npm verb cli   'yarn',
npm verb cli   '--verbose' ]
npm info using npm@6.14.4
npm info using node@v10.19.0
npm verb npm-session 4403626286b4d9ec
npm http fetch GET 200 https://registry.npmjs.org/yarn 208ms
npm http fetch GET 200 https://registry.npmjs.org/yarn/-/yarn-1.22.8.tgz 199ms
npm timing stage:loadCurrentTree Completed in 452ms
npm timing stage:loadIdealTree:cloneCurrentTree Completed in 0ms
npm timing stage:loadIdealTree:loadShrinkwrap Completed in 1ms
npm timing stage:loadIdealTree:loadAllDepsIntoIdealTree Completed in 6ms
npm timing stage:loadIdealTree Completed in 10ms
npm timing stage:generateActionsToTake Completed in 5ms
npm verb correctMkdir /root/.npm/_locks correctMkdir not in flight; initializing
npm verb lock using /root/.npm/_locks/staging-3a08f0df5026584d.lock for /usr/local/lib/node_modules/.staging
npm timing action:extract Completed in 106ms
npm timing action:finalize Completed in 3ms
npm timing action:refresh-package-json Completed in 15ms
npm info lifecycle yarn@1.22.8~preinstall: yarn@1.22.8

> yarn@1.22.8 preinstall /usr/local/lib/node_modules/yarn
> :; (node ./preinstall.js >& /dev/null || true)


root@bee39e307c36:/tmp# echo $?
1
root@bee39e307c36:/tmp# npm --version
6.14.4
```

I'm on Docker with ubuntu focal version

And maybe if command `npm install -g yarn` will not work on root since version > 1.22.5, version must be 2.x.x ?

**Comment by arcanis** [comment]
I've released **1.22.10** that avoids file descriptors altogether. Can you confirm it works on your side?

> @arcanis Can this be fixed real soon (v1.22.9)? It is blocking our work…

For future reference, you can (should) pin your version to any previous one (by using `yarn@1.22.6`, etc), which is a good practice anyway regardless of the conditions. You can also use the [`yarn-path` setting](https://classic.yarnpkg.com/en/docs/cli/policies) to ensure that upgrades go through the appropriate review processes.

**Comment by gjvoosten** [comment]
1.22.10 works like a charm, @arcanis ; thanks for the speedy response!

**Comment by dcombrouse** [comment]
Don't work for me, but I'm going to use 1.22.5 version

**Comment by arcanis** [comment]
That sounds good - if you have an idea why is the execution failing, feel free to open another issue (to avoid us spamming other people from this thread). Without logs however, I won't be able to help much.

Fwiw we don't plan to add any more features to Yarn 1, as all of our resources have shifted to Yarn 2. The past few commits have been aimed toward making the transition a bit easier, in particular thanks to the [Corepack](https://github.com/arcanis/corepack) initiative which we hope will make it easier to use Yarn (both 1 & 2) by removing the need to manually install them.

In the meantime, sorry for the disruption. I hope things will work fine now 🙂

**Comment by bjentsch** [comment]
There's an issue with this on Linux; the preinstall step errors out when executed:
```
��� YN0000: ��� yarn@npm:1.22.10 STDERR SyntaxError: Expected "$", "$(", "$((", "${", "'", "\"", "\\", or [ \t] but "&" found (line 1, column 40).
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at fe (/var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:55:361665)
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at parse (/var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:55:376698)
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at i (/var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:2:456775)
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at T (/var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:2:488855)
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at i.configuration.reduceHook.script (/var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:2:367760)
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at /var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:2:367882
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at async a.mktempPromise (/var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:2:447806)
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at async Module.k (/var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:2:367500)
��� YN0000: ��� yarn@npm:1.22.10 STDERR     at async /var/lib/jenkins/workspace/frontend/.yarn/releases/yarn-2.2.2.js:2:324596
��� YN0009: ��� yarn@npm:1.22.10 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-fe2baaa4/build.log)
��� YN0009: ��� yarn@npm:1.22.10 couldn't be built successfully (exit code 1, logs can be found here: /tmp/xfs-fe2baaa4/build.log)
```
We've reverted our yarn version to `1.22.7` where this issue doesn't exist yet.

**Comment by gengjiawen** [comment]
> Don't work for me, but I'm going to use 1.22.5 version

Looks like 1.22.10 works on Node.js 14, not 10 from what I test.

**Comment by gengjiawen** [comment]
> > Don't work for me, but I'm going to use 1.22.5 version
> 
> Looks like 1.22.10 works on Node.js 14, not 10 from what I test.

Interesting, you may need to update npm too. `6.14.4` not works, `6.14.8` works.

**Comment by arcanis** [comment]
@benjie This parsing issue is unique to Yarn 2. Can you open a thread on [our repository](https://github.com/yarnpkg/berry)?

**Comment by benjie** [comment]
@arcanis I'm not sure why I was @-d on this? Perhaps you meant to reply to a different issue?

**Comment by arcanis** [comment]
Sorry, I meant @bjentsch 😶

**Comment by bjentsch** [comment]
> This parsing issue is unique to Yarn 2. Can you open a thread on our repository?
Will do @arcanis :+1:

**Comment by samal-rasmussen** [comment]
> I've just released the 1.22.8 which uses the following script; it should prevent permission errors (since the script aims to make a future migration easier it's not a critical problem if it can't run), at the cost of the script not running anymore on Windows. I'll update the changelog.
> 
> ```
> :; (node ./preinstall.js >& /dev/null || true)
> ```

I'm having trouble parsing this. Does this mean Windows should still work if you don't install as admin? Or are you saying you're dropping support for Windows?

**Comment by arcanis** [comment]
The postinstall script simply won't have any effect. The rest of Yarn (ie, everything you care about) will work.

**Comment by edalongeville** [comment]
This is still happening to me.

Steps to reproduce:
```
docker run -ti --entrypoint /bin/bash ubuntu
```
Inside the container:
```
apt-get update
apt-get install npm
npm install -g yarn
```
This will work:
```
npm install -g yarn@1.22.6
```

**Comment by ScottPierce** [comment]
I'm seeing this fail for docker ubuntu containers:

```
FROM ubuntu:20.04

USER root

RUN apt update \
    && apt upgrade -y

RUN apt install -y --no-install-recommends \
    curl \
    && apt-get clean \
    && rm -rf /var/lib/apt/lists/*

RUN apt update \
    && apt upgrade -y

RUN apt install -y --no-install-recommends \
    nodejs \
    npm \
    openjdk-8-jdk \
    && apt clean \
    && rm -rf /var/lib/apt/lists/*

RUN npm install -g yarn \
    && yarn global add firebase-tools@8.12.1 \
    && yarn cache clean \
    && firebase setup:emulators:firestore
```

**Comment by malisoft** [comment]
> This is still happening to me.
> 
> Steps to reproduce:
> 
> ```
> docker run -ti --entrypoint /bin/bash ubuntu
> ```
> 
> Inside the container:
> 
> ```
> apt-get update
> apt-get install npm
> npm install -g yarn
> ```
> 
> This will work:
> 
> ```
> npm install -g yarn@1.22.6
> ```

try run this command: sudo npm install -g npm 
it's works for me

**Comment by s10z** [comment]
I faced the same issue with a fresh Ubuntu 20.04.01 LTS installation. First the `npm install -g yarn` failed due to permission errors, leveraging it to `sudo npm install -g yarn` led to a 
``; (node ./preinstall.js > /dev/null 2>&1 || true)`` 
response but no installation. Finally the `sudo npm install -g npm` before `sudo npm install -g yarn` as  @malisoft pointed out fixed it for me to have yarn 1.22.10 installed without additional fiddling.

**Comment by camilojm27** [comment]
> I faced the same issue with a fresh Ubuntu 20.04.01 LTS installation. First the `npm install -g yarn` failed due to permission errors, leveraging it to `sudo npm install -g yarn` led to a
> `; (node ./preinstall.js > /dev/null 2>&1 || true)`
> response but no installation. Finally the `sudo npm install -g npm` before `sudo npm install -g yarn` as @malisoft pointed out fixed it for me to have yarn 1.22.10 installed without additional fiddling.

Thank you, this worked for me

**Comment by Larpoux** [comment]
> try run this command: **sudo npm install -g npm**
> and after: **sudo npm install -g yarn**
> it's works for me:
> 
> > `malisoftdev@malisoftdev ~ [1]> sudo npm install -g npm`
> > /usr/local/bin/npm -> /usr/local/lib/node_modules/npm/bin/npm-cli.js
> > /usr/local/bin/npx -> /usr/local/lib/node_modules/npm/bin/npx-cli.js
> > 
> > * npm@6.14.9
> >   added 435 packages from 887 contributors in 11.802s
> 
> > `malisoftdev@malisoftdev ~> sudo npm install -g yarn`
> 
> > yarn@1.22.10 preinstall /usr/local/lib/node_modules/yarn
> > :; (node ./preinstall.js > /dev/null 2>&1 || true)
> 
> > /usr/local/bin/yarn -> /usr/local/lib/node_modules/yarn/bin/yarn.js
> > /usr/local/bin/yarnpkg -> /usr/local/lib/node_modules/yarn/bin/yarn.js
> > 
> > * yarn@1.22.10
> >   updated 1 package in 2.415s
> 
> > `malisoftdev@malisoftdev ~>`

Thank you @malisoft : You saved me. I tried many things and your solution is the only it worked for me.
